JetNews/app/src/main/java/com/example/jetnews/ui/interests/InterestsViewModel.kt

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+package com.example.jetnews.ui.interests
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.example.jetnews.data.interests.InterestsRepository
+import com.example.jetnews.data.interests.TopicSelection
+import com.example.jetnews.data.interests.TopicsMap
+import com.example.jetnews.data.successOr
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+/**
+ * UI state for the Interests screen
+ */
+data class InterestsUiState(
+    val topics: TopicsMap = emptyMap(),
+    val people: List<String> = emptyList(),
+    val publications: List<String> = emptyList(),
+    val loading: Boolean = false,
+)
+
+class InterestsViewModel(
+    private val interestsRepository: InterestsRepository
+) : ViewModel() {
+
+    // UI state exposed to the UI
+    private val _uiState = MutableStateFlow(InterestsUiState(loading = true))
+    val uiState: StateFlow<InterestsUiState> = _uiState.asStateFlow()
+
+    val selectedTopics =
+        interestsRepository.observeTopicsSelected().stateIn(
+            viewModelScope,
+            SharingStarted.WhileSubscribed(5000),
+            emptySet()
+        )
+
+    val selectedPeople =
+        interestsRepository.observePeopleSelected().stateIn(
+            viewModelScope,
+            SharingStarted.WhileSubscribed(5000),
+            emptySet()
+        )
+
+    val selectedPublications =
+        interestsRepository.observePublicationSelected().stateIn(
+            viewModelScope,
+            SharingStarted.WhileSubscribed(5000),
+            emptySet()
+        )
+
+    init {
+        refreshAll()
+    }
+
+    fun toggleTopicSelection(topic: TopicSelection) {
+        viewModelScope.launch {
+            interestsRepository.toggleTopicSelection(topic)
+        }
+    }
+
+    fun togglePersonSelected(person: String) {
+        viewModelScope.launch {
+            interestsRepository.togglePersonSelected(person)
+        }
+    }
+
+    fun togglePublicationSelected(publication: String) {
+        viewModelScope.launch {
+            interestsRepository.togglePublicationSelected(publication)
+        }
+    }
+
+    /**
+     * Refresh topics, people, and publications
+     */
+    private fun refreshAll() {
+        _uiState.update { it.copy(loading = true) }
+
+        viewModelScope.launch {
+            val topics = async { interestsRepository.getTopics() }
+            val people = async { interestsRepository.getPeople() }
+            val publications = async { interestsRepository.getPublications() }
+
+            // An incomplete UI state is created to not `await` in the _uiState.update lambda
```

Comment:
How is this incomplete if you're awaiting every type of data?

Reply:
Loading is not there